pydantic/_internal/_discriminated_union.py

```diff
@@ -285,17 +279,14 @@ def _handle_choice(self, choice: core_schema.CoreSchema) -> None:
             # Reverse the choices list before extending the stack so that they get handled in the order they occur
             choices_schemas = [v[0] if isinstance(v, tuple) else v for v in choice['choices'][::-1]]
             self._choices_to_handle.extend(choices_schemas)
-        elif choice['type'] == 'definition-ref':
-            if choice['schema_ref'] not in self.definitions:
-                raise MissingDefinitionForUnionRef(choice['schema_ref'])
-            self._handle_choice(self.definitions[choice['schema_ref']])
```

Comment:
Important!!! If a choice is of type `definition-ref`, we want to just reuse that ref for the given choice. Before, we were going through this whole thing of fetching the value from definitions, then using that, but that ends up not working for nested / recursive schemas.

Our schema walking logic walks through both the schema and the definitions, so we can rest easy knowing that unions will be converted to tagged unions in the definitions list as well.

Reply:
Does this work with JSON schema generation if e.g. the ref'ed schema is itself a discriminated union? Maybe that can't happen, and either way this seems like an improvement if no tests fail, but still

Reply:
Good question. I believe that the walk core schema logic handles definitions schemas such that the function being applied during the walk is also applied to all of the definitions in a definitions schema, so that's why I felt comfortable removing this step.

This can be seen via the example test I added - discriminated union transformation logic is applied to the 2 schemas in the definitions list that require said changes!

Reply:
Good question. I believe that the walk core schema logic handles definitions schemas such that the function being applied during the walk is also applied to all of the definitions in a definitions schema, so that's why I felt comfortable removing this step.

This can be seen via the example test I added - discriminated union transformation logic is applied to the 2 schemas in the definitions list that require said changes!